### DISCUSSION TITLE
How do I define and query a read / view function in a Move module?

### DISCUSSION THREAD

**Discussion by banool** [aptos-labs/aptos-developer-discussions#38] [Questions] [answered]
Two questions:
1. Is it possible to use the Python SDK to query read/write functions in a Move module?
2. How do you define a read function in a Move module?

**Comment by banool** [comment, marked as answer]
Aptos has support for view functions. This is a view function in the `coin` module in the framework:
```
#[view]
/// Returns the balance of `owner` for provided `CoinType`.
public fun balance<CoinType>(owner: address): u64 acquires CoinStore {
    assert!(
        is_account_registered<CoinType>(owner),
        error::not_found(ECOIN_STORE_NOT_PUBLISHED),
    );
    borrow_global<CoinStore<CoinType>>(owner).coin.value
}
```
From [here](https://github.com/aptos-labs/aptos-core/blob/c7375b86ff492862f49ee56e994fed9c8af71709/aptos-move/framework/aptos-framework/sources/coin.move#L252-L260).

You can query it like this with the Python SDK:
```python
client = RestClient("https://api.mainnet.aptoslabs.com/v1")
func = "0x1::coin::balance"
type_arguments = ["0x1::aptos_coin::AptosCoin"]
arguments = ["0x123"]
await client.view(func, type_arguments, arguments)
```

Here is how you do it with the TS SDK:
```typescript
const config = new AptosConfig({ network: Network.MAINNET });
const aptos = Aptos(config);
const result = await aptos.view({
    payload: {
        function: "0x1::coin::balance",
        typeArguments: ["0x1::aptos_coin::AptosCoin]",
        functionArguments: ["0x123"],
    },
});
```

Here is how you do it with the CLI:
```bash
aptos move view --url https://api.mainnet.aptoslabs.com --function-id 0x1::coin::balance --type-args '0x1::aptos_coin::AptosCoin' --args address:0x123
```

**Reply by karamelniycoder** [reply]
dear @banool can u explain where u found method "view" in RestClient?

> await client.view(func, type_arguments, arguments)

python sdk is clearly empty: https://github.com/search?q=repo%3Aaptos-labs%2Faptos-python-sdk+view&type=code

**Reply by gregnazario** [reply]
Haven't migrated to the new repository yet, so I'll hide it for now.

See here: https://github.com/aptos-labs/aptos-core/blob/main/ecosystem/python/sdk/aptos_sdk/async_client.py#L785-L822
